StashiKu/aws-shop-be
# Aws-BE-6: SQS and SNS Async Microservices Communication

### link to FE deploy - https://d3ar6ba8glakkn.cloudfront.net/
### link to PR for FE part - https://github.com/StashiKu/shop-angular-cloudfront/pull/5

  ### Functionality:
 - **migrated** from `DynamoDB` to `PostgreSQL`
  -  `importFileParser ` - uses a readable stream to get an object from S3, parses it using csv-parser package and creates SQS message with product data
  - `catalogBatchProcess` - gets message from SQS, creates product and sends notification with SNS

### How to test
 1. You can take this file [test-task-6.csv](https://github.com/StashiKu/aws-shop-be/files/9949370/test-task-6.csv).
 2. Upload it via admin Ui https://d3ar6ba8glakkn.cloudfront.net/admin/products
 3. Reload page - you should see new product
 4. Ping me and I will send you screenshot of notifications.

  ### More info:
 -  PostgreSQL 
 - `Jest` is used for testing. Run `npm run test` to run tests
 - `TypeScript`
 - `eslint`
 
**link to requirements** https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/6_async_microservices_communication/task.md

### Evaluation criteria
[**1**] - File serverless.yml contains configuration for catalogBatchProcess function
[**2**] - File serverless.yml contains policies to allow lambda catalogBatchProcess function to interact with SNS and SQS
[**3**] - File serverless.yml contains configuration for SQS catalogItemsQueue
[**4**] - File serverless.yml contains configuration for SNS Topic createProductTopic and email subscription

**Additional (optional) tasks**
[**+1**] -  set a Filter Policy for SNS `createProductTopic` in serverless.yml and create an additional email subscription to distribute messages to different emails depending on the filter for any product attribute
<img width="473" alt="Screen Shot 2022-11-07 at 11 23 55 AM" src="https://user-images.githubusercontent.com/26601879/200261239-137c42fe-7dfc-45a8-b6a4-6cee3dd9bc99.png">

[**+1**] - `catalogBatchProcess` lambda is covered 
by unit tests https://github.com/StashiKu/aws-shop-be/pull/4/files#diff-f344aab36d2fe3aa5c6a24025fde3a5db08c79661e618bf036eb1e89b198924a

### Total score - 6

**Thanks for reviewing** 👍 

### ⚠️ PS: if approve, please write a comment below "Approved"
